src/node_native_module.cc

```diff
@@ -0,0 +1,325 @@
+  EscapableHandleScope scope(env->isolate());
+  Local<Context> context = env->context();
+  Isolate* isolate = env->isolate();
+
+  Local<Value> result =
+      env->native_modules_source()->Get(context, id).ToLocalChecked();
+  CHECK(result->IsString());
+  Local<String> source = result.As<String>();
+
+  Local<String> filename =
+      String::Concat(isolate, id, FIXED_ONE_BYTE_STRING(isolate, ".js"));
+  Local<Integer> line_offset = env->native_module_line_offset();
+  Local<Integer> column_offset = env->native_module_column_offset();
+  ScriptOrigin origin(filename, line_offset, column_offset);
+
+  bool use_cache = false;
+  ScriptCompiler::CachedData* cached_data = nullptr;
+
+  // 1. We won't even check the existence of the cache if the binary is not
+  //    built with them.
+  // 2. If we are generating code cache for tools/general_code_cache.js, we
+  //    are not going to use any cache ourselves.
+  if (native_module_has_code_cache && !produce_code_cache) {
+    cached_data = GetCachedData(env, id);
+    if (cached_data != nullptr) {
+      use_cache = true;
+    }
+  }
+
+  ScriptCompiler::Source script_source(source, origin, cached_data);
+
+  ScriptCompiler::CompileOptions options;
+  if (produce_code_cache) {
+    options = ScriptCompiler::kEagerCompile;
+  } else if (use_cache) {
+    options = ScriptCompiler::kConsumeCodeCache;
+  } else {
+    options = ScriptCompiler::kNoCompileOptions;
+  }
+
+  MaybeLocal<Function> maybe_fun =
+      ScriptCompiler::CompileFunctionInContext(context,
+                                               &script_source,
+                                               parameters_count,
+                                               parameters,
+                                               0,
+                                               nullptr,
+                                               options);
+
+  TryCatch try_catch(isolate);
+  Local<Function> fun;
+  // This could fail when there are early errors in the native modules,
+  // e.g. the syntax errors
+  if (maybe_fun.IsEmpty() || !maybe_fun.ToLocal(&fun)) {
+    DecorateErrorStack(env, try_catch);
+    try_catch.ReThrow();
+    return scope.Escape(Local<Value>());
+  }
+
+  if (use_cache) {
+    // If the cache is rejected, something must be wrong with the build
+    // and we should just crash.
+    CHECK(!script_source.GetCachedData()->rejected);
+    if (env->native_modules_with_cache()->Add(context, id).IsEmpty()) {
+      return scope.Escape(Local<Value>());
+    }
+  } else {
+    if (env->native_modules_without_cache()->Add(context, id).IsEmpty()) {
+      return scope.Escape(Local<Value>());
+    }
+  }
+
+  if (produce_code_cache) {
+    std::unique_ptr<ScriptCompiler::CachedData> cached_data(
+        ScriptCompiler::CreateCodeCacheForFunction(fun));
+    CHECK_NE(cached_data, nullptr);
+    char* data =
+        reinterpret_cast<char*>(const_cast<uint8_t*>(cached_data->data));
```

Comment:
I don’t think you need to cast the `const` away?

Reply:
I do...otherwise the compiler barks at me..

Reply:
Hm … my compiler (gcc 7.3.0) accepts this:

```diff
diff --git a/src/node_native_module.cc b/src/node_native_module.cc
index 50f9e3dc3a53..d6b938309d5c 100644
--- a/src/node_native_module.cc
+++ b/src/node_native_module.cc
@@ -262,8 +262,7 @@ Local<Value> NativeModule::Compile(Environment* env,
     std::unique_ptr<ScriptCompiler::CachedData> cached_data(
         ScriptCompiler::CreateCodeCacheForFunction(fun));
     CHECK_NE(cached_data, nullptr);
-    char* data =
-        reinterpret_cast<char*>(const_cast<uint8_t*>(cached_data->data));
+    const char* data = reinterpret_cast<const char*>(cached_data->data);
     // Since we have no API to create a buffer from a new'ed pointer,
     // we will need to copy it - but this code path is only run by the
     // tooling that generates the code cache to be bundled in the binary
```

So it sounds like a compiler bug? Maybe we could leave a comment so that people aren’t tempted to refactor the `const_cast` away?

Reply:
Just FTR, mutating the value of cast-away-const is undefined behaviour, and in recent MSVS version, it will probably crash (definatly crash in a Debug build).

Reply:
There is no mutation of the underlying data – the `const_cast` is valid, if it is necessary.

Reply:
If `const_cast` actually causes crashes when its valid, this is probably one of the last place we need to worry about :) (it's only invoked by code cache generator, which hopefully no-one but us uses)